url4aac7e6f55_simon0191_tetris3d_tgz-pJ8-Tetris3dJ8/src/vsdk/toolkit/io/image/ImagePersistence.java

```diff
@@ -0,0 +1,693 @@
+              catch ( Exception e ) {
+                  VSDK.reportMessage(null, VSDK.ERROR, "importRGB (B)",
+                                     "Cannot import image file \"" + inImageFd.getAbsolutePath() + "\"");
+		  e.printStackTrace();
+                 throw new ImageNotRecognizedException("Error reading internal file:\n" + e, inImageFd);
+            }
+        }
+        throw new ImageNotRecognizedException("Image not recognized", inImageFd);
+    }
+
+    private static void transferPixels(int[] ori, byte[] dest, int w, int h, int pixelDepth)
+    {
+        int bPos=0;
+        for(int i=0; i<w*h; i++)
+        {
+            dest[bPos]=(byte)(ori[i]>>16);
+            bPos++;
+            dest[bPos]=(byte)(ori[i]>>8);
+            bPos++;
+            dest[bPos]=(byte)(ori[i]>>0);
+            bPos++;
+            if(pixelDepth==32)
+            {
+                dest[bPos]=(byte)(ori[i]>>24);
+                bPos++;
+            }
+        }
+    }
+    
+    /**
+    This method writes the contents of the specified image to a file in 
+    binary JPEG image format. Returns true if everything
+    works fine, false if something fails, like a permission access denied
+    or if storage device runs out of space.
+    */    
+    public static boolean exportJPG(File fd, Image img)
+    {
+        try {
+            FileOutputStream fos = new FileOutputStream(fd);
+
+            exportJPG(fos, img);
+
+            fos.close();
+        }
+        catch ( Exception e ) {
+            return false;
+        }
+        return true;
+    }
+
+    public static void exportJPG(OutputStream os, Image img)
+        throws Exception
+    {
+        BufferedImage bimg;
+        int x, y, xSize, ySize;
+        RGBPixel p;
+
+        xSize = img.getXSize();
+        ySize = img.getYSize();
+        bimg =  new BufferedImage(xSize, ySize, BufferedImage.TYPE_INT_RGB);
+        for ( y = 0; y < ySize; y++ ) {
+            for ( x = 0; x < xSize; x++ ) {
+                p = img.getPixelRgb(x, y);
+                bimg.setRGB(x, y, 
+                  (VSDK.signedByte2unsignedInteger(p.r)) * 256 * 256 +
+                  (VSDK.signedByte2unsignedInteger(p.g)) * 256 +
+                  (VSDK.signedByte2unsignedInteger(p.b))
+                );
+            }
+        }
+
+        ImageIO.write(bimg, "jpg", os);
+
+        // OLD DEPRECATED API, DO NOT USE!
+        //com.sun.image.codec.jpeg.JPEGImageEncoder jpeg;
+        //jpeg = com.sun.image.codec.jpeg.JPEGCodec.createJPEGEncoder(fos);
+        //jpeg.encode(bimg);
+    }
+
+    /**
+    This method writes the contents of the specified image to a file in 
+    binary JPEG image format. Returns true if everything
+    works fine, false if something fails, like a permission access denied
+    or if storage device runs out of space.
+    */    
+    public static boolean exportPNG(File fd, Image img)
+    {
+        try {
+            FileOutputStream fos = new FileOutputStream(fd);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *fos*. The resource is referenced at line 361. The resource is closed at line 363. There are other execution paths that do not contain closure statements, for example, when ImagePersistence.exportPNG throws an exception. Ensure *fos* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).